usr/src/cmd/smbsrv/smbd/smbd_logon.c

```diff
@@ -109,7 +111,20 @@ smbd_user_auth_logon(smb_logon_t *user_info)
 	} else {
 		tmp_user.lg_e_username = tmp_user.lg_username;
 	}
-	tmp_user.lg_e_domain = tmp_user.lg_domain;
+
+	/* Handle user@domain format. */
+	if (tmp_user.lg_domain[0] == '\0' &&
+	    (p = strchr(tmp_user.lg_e_username, '@')) != NULL) {
+		buf = strdup(tmp_user.lg_e_username);
+		if (buf == NULL)
+			goto errout;
+		p = buf + (p - tmp_user.lg_e_username);
```

Comment:
we need a strtok() in the kernel. I'm assuming we don't have one already.

Reply:
I can live with it as it is.

Reply:
http://src.illumos.org/source/xref/illumos-gate/usr/src/common/util/string.c#714

Reply:
Oh wait, this is user space.  Maybe just use strtok_r() instead of baking your own.

Reply:
I don't think strtok is the 'right tool' here. We already know, from strchr(), where in the string the character we want to overwrite is; Converting just the if-block to use strtok would just make us search the string for it again, and I don't see why we'd want to do that. And since we need to dup the string if we're going to modify it, we can't use strtok() to find the token in the first instance. I'd rather keep it as-is.